crates/ecs/src/lib.rs

```diff
@@ -331,22 +395,42 @@ impl<'system> SystemExecutionGuard<'system> {
 
 /// Schedules systems in no particular order, with no regard to dependencies.
 #[derive(Default, Debug)]
-pub struct Unordered<'systems>(&'systems [Box<dyn System>]);
+pub struct Unordered<'systems> {
+    systems: &'systems [Box<dyn System>],
+    wait_until_next_call: bool,
+}
 
 impl<'systems> Schedule<'systems> for Unordered<'systems> {
     fn generate(systems: &'systems [Box<dyn System>]) -> ScheduleResult<Self> {
-        Ok(Self(systems))
+        Ok(Self {
+            systems,
+            wait_until_next_call: false,
+        })
     }
 
-    fn currently_executable_systems(
+    fn currently_executable_systems_with_reaction(
         &mut self,
+        new_tick_reaction: NewTickReaction,
     ) -> ScheduleResult<Vec<SystemExecutionGuard<'systems>>> {
-        Ok(self
-            .0
+        let all_systems = self
+            .systems
             .iter()
             .map(|system| system.as_ref())
             .map(|system| SystemExecutionGuard::create(system).0)
-            .collect())
+            .collect();
+
+        if new_tick_reaction == NewTickReaction::ReturnError && self.wait_until_next_call {
+            // Make sure next call to this function will return systems.
+            self.wait_until_next_call = false;
+            Err(ScheduleError::NewTick)
```

Comment:
Might not be a real suggestion.  I´m just trying to understand the code better.

Like in this case here there seems to be some Err being passed around between function calls even though it, from my understanding, seems to be intended behaviour. I still find it diffucult for errors being actually requested behaviour from systems but this may just be what is usually done in rust. Trying to understand why this would be the case. Is it that Err passes flags while Ok handles actual data? 

Are there ways to rename the Result type to not use Err and instead use for example a name like Flag if uses like this is indeed intended behaviour. This could be more trouble than it's worth though.

Reply:
It depends on how you look at it, I suppose. The way I saw it was that it's an error because when it's returned the function doesn't do what it's meant to do, which is to return the next batch of available systems. If `new_tick_reaction` is set to `NewTickReaction::ReturnNewTick` then the function behaves as expected. 

Though, you could argue that since the caller is explicitly requesting for `NewTickReaction::ReturnError` then the function _is_ doing what it's meant to do. 

The reason I opted for using an error instead of making the normal success-value an enum with two variants (one `NewTick` and one `NextSystems(...)`) is because that'd mean every caller of the function would need to unwrap that enum and handle both cases every time it's used. For example, these usages would not work if `currently_executable_systems` do not return an iterable:
https://github.com/martinjonsson01/recs/blob/1c56bbe2a927b060143e5d34461602288f2accb5/crates/ecs/src/lib.rs#L264-L279
By using an error, most callers can simply ignore it using `?` and only those special-case callers who need to be aware of when a new tick has happened can handle it like this:
https://github.com/martinjonsson01/recs/blob/1c56bbe2a927b060143e5d34461602288f2accb5/crates/scheduler/src/executor/mod.rs#L253-L270

So I would argue that in the first example above, in lib.rs, it is actually unexpected behavior for the error `ScheduleError::NewTick` to appear since it hasn't been requested. On the other hand in the second example, in executor/mod.rs, that "error" is handled explicitly as something that is expected. 

I agree that in some situations it looks a bit weird for this to be an error, but it turned out to be the simplest way of adding this extra level of behavior that I could come up with at the time. I'm open to any suggestions that avoid using an error while still preserving the "common-case" functionality of simply iterating over the returned value.